### PR TITLE
Issue fix 11688

### PR DESCRIPTION
**Pull request by zaverichintan**
By this code we're making wrapper functions for setting and getting cookies from array.

Signed-off-by:Chintan zaveri zaverichintan5@gmail.com

### CODE REVIEW

**Comment by zaverichintan** [comment]
@zixtor kindly help in Over view

File: libraries/plugins/auth/AuthenticationCookie.php [hunk 1]
```diff
@@ -405,14 +405,14 @@ public function authCheck()
         }
 
         // check cookies
-        if (empty($_COOKIE['pmaUser-' . $GLOBALS['server']])
-            || empty($_COOKIE['pma_iv-' . $GLOBALS['server']])
+        if (getCookieArr('pmaUser','pmaUser-' . $GLOBALS['server'])== null
```
Comment (zixtor) [inline libraries/plugins/auth/AuthenticationCookie.php]: PHP Fatal error:  Call to undefined function libraries\plugins\auth\getCookieArr() in libraries/plugins/auth/AuthenticationCookie.php on line 408

File: libraries/plugins/auth/AuthenticationCookie.php [hunk 2]
```diff
@@ -405,14 +405,14 @@ public function authCheck()
...
+            || getCookieArr('pmaAuth','pma_iv-' . $GLOBALS['server'])==null
         ) {
             return false;
         }
 
         $GLOBALS['PHP_AUTH_USER'] = $this->cookieDecrypt(
-            $_COOKIE['pmaUser-' . $GLOBALS['server']],
+            $getCookieArr('pmaUser','pmaUser-' . $GLOBALS['server']),
```
Comment (zixtor) [inline libraries/plugins/auth/AuthenticationCookie.php]: typo: $getCookieArr

File: libraries/Config.php [hunk 3]
```diff
@@ -1868,7 +1875,85 @@ public function setCookie($cookie, $value, $default = null,
...
+    /**
+     * sets the cookie
+     * @param string      $cookieArr Cookie array name
+     * @param DocBlockr   $cookie    sub cookie name
+     * @param mixed       $value     sub cookie value
+     * @param string      $default  default value
+     * @param int         $validity validity of cookie in seconds (default is one month)
+     * @param bool        $httponly whether cookie is only for HTTP (and not for scripts)
+     */
+    public function setCookieArr($cookieArr,$cookie, $value, $default = null,
```
Comment (zixtor) [inline libraries/Config.php]: setCookieArr seems to duplicate the logic in setCookie, which should not be necessary.

**Comment by zixtor** [comment]
I was not able to test the PR.
As for qualitative help, in general, volunteers will readily review what you wish to get merged and may also provide some suggestions in flow with you own work direction; but beyond that it poses limitations and concerns to provide thorough help, more so if you wish to apply for GSoC, as it will violate fair play for other candidates.

**Comment by ibennetch** [comment]
Hi, when making changes to a pull request you don't need to close it to start a new one; you can just push in to your repository and the pull request will automatically be updated. Sometimes you'll have to `--force` the push, such as if you squash the commit history, but this still generally works.

**Comment by zaverichintan** [comment]
How do I test the three types of authentication on local? 
And should I import the libraries/config.php in authentication.php?

**Comment by zixtor** [comment]
Please refer to https://wiki.phpmyadmin.net/pma/Auth_types. 
I don't understand the second question about "import", but you can simply do $GLOBALS['PMA_Config']->getCookie(...) if that is what you meant.

File: libraries/Config.php [hunk 4]
```diff
@@ -20,6 +20,13 @@
 $GLOBALS['pma_config_loading'] = false;
 
 /**
+ * Array to store information temporarily and later set cookie(see setcookie function).
+ */
+$GLOBALS["pmaUser"] = array();
+$GLOBALS["pmaAuth"] = array();
+$GLOBALS["pmaTest"] = array();
```
Comment (zixtor) [inline libraries/Config.php]: Why do you need these globals, just for local use?

File: libraries/plugins/auth/AuthenticationCookie.php (hunk 1 above)
Comment (zaverichintan) [inline libraries/plugins/auth/AuthenticationCookie.php]: @zixtor To use getCookieArr in libraries/plugins/auth/AuthenticationCookie.php how should I include the config file (getCookieArr Function) .
Kindly suggest the method .

File: libraries/plugins/auth/AuthenticationCookie.php (hunk 2 above)
Comment (zaverichintan) [inline libraries/plugins/auth/AuthenticationCookie.php]: Fixed

File: libraries/Config.php (hunk 3 above)
Comment (zaverichintan) [inline libraries/Config.php]: I agree with you , but this is done as setCookieArr function may replace setCookie function in future.

File: libraries/Config.php (hunk 4 above)
Comment (zaverichintan) [inline libraries/Config.php]: The use of cookies didn't had wrapper function for retrieving cookies, so by making it global array , we can access it anywhere.

File: libraries/plugins/auth/AuthenticationCookie.php (hunk 1 above)
Comment (zixtor) [inline libraries/plugins/auth/AuthenticationCookie.php]: @zaverichintan Please see my previous comment https://github.com/phpmyadmin/phpmyadmin/pull/11921#issuecomment-180753203, I did answer it.

**Comment by zaverichintan** [comment]
should I include the config.php in authenticationcookie?
Kindly help me

File: libraries/plugins/auth/AuthenticationCookie.php [hunk 5]
```diff
@@ -31,6 +31,10 @@
 require './libraries/plugins/auth/swekey/swekey.auth.lib.php';
 
 /**
+ * To use getCookieArr function
+ */
+require './libraries/Config.php';
```
Comment (nijel) [inline libraries/plugins/auth/AuthenticationCookie.php]: At most it should be require_once. But I think it should be rather used through autoloading...

File: libraries/Config.php [hunk 6]
```diff
@@ -1868,8 +1875,81 @@ public function setCookie($cookie, $value, $default = null,
...
+
+                return setcookie(
+                    $cookieArr,
+                    $json_encoded,
+                    $validity,
+                    $this->getCookiePath(),
+                    '',
+                    $this->isHttps(),
+                    $httponly
+                );
```
Comment (nijel) [inline libraries/Config.php]: Why don't you either replace existing setCookie or use it here? Copying the code has never done anything good.

File: libraries/plugins/auth/AuthenticationCookie.php (hunk 5 above)
Comment (zaverichintan) [inline libraries/plugins/auth/AuthenticationCookie.php]: Should i make a change in autoloader ?

File: libraries/Config.php (hunk 6 above)
Comment (zaverichintan) [inline libraries/Config.php]: I made two functions right now so other developers can easily use the existing function and also to solve cookie related errors.

Comment (poush) [inline libraries/Config.php]: I too believe that it is a bad idea to duplicate the code. And after all phpmyadmin codebase should be more improved to OO not to make it more procedural

Comment (zaverichintan) [inline libraries/Config.php]: I agree with you , but this code is a copy to replace the function once everyone agrees to the array json concept.

**Comment by poush** [comment]
One my first look there many places where $_COOKIE is directly used and to make it work i think there is need to update those lines like ```isset($_COOKIE['...']``` , etc from whole codebase.

**Comment by zaverichintan** [comment]
@poush I am also thinking the same but the cookies are set when http header is sent, I only replaced the usage with the wrapper function already used. 
Replacing leads to different issues.

**Comment by poush** [comment]
But then how it will work if half of the code expecting old style of cookies when half portion is setting up cookies in JSON

**Comment by ibennetch** [comment]
Near the bottom of the CI report we can see the error output:

> [exec] PHP Fatal error:  Cannot declare class PMA\libraries\Config, because the name is already in use in /home/travis/build/phpmyadmin/phpmyadmin/libraries/Config.php on line 34

Which I haven't investigated fully, but am guessing is related to `+require './libraries/Config.php';` which @nijel has already mentioned above. Try fixing that and then check the test again.

**Comment by zaverichintan** [comment]
@ibennetch thank you for your response, trying that. 
@poush there are different parameters which are same, only one parameter has changed (value of the cookie is now saved as json)

**Comment by poush** [comment]
Did you tested your code ? Because what i am saying is lets suppose you have

```[pmaConfig] => [pma_lang,pma_fontSize,....]```

but in LanguageManager.php
```
  725:         if (! empty($_COOKIE['pma_lang'])) {
  726:             if (isset($langs[$_COOKIE['pma_lang']])) {
  727:                 return $langs[$_COOKIE['pma_lang']];
  728              }
```
so it expects the old way

and about this
```
/phpmyadmin/import_status.php:
   45  
   46      session_name('phpMyAdmin');
   47:     session_id($_COOKIE['phpMyAdmin']);
   48  }
   49   */
```

Above that, whenever you call setCookieArr(), a new "set-cookie" header will be made. Having multiple header for same name,path has undefined behaviour. For example my chrome and safari is saving first one only while firefox is saving N times the first one header.

**Comment by zaverichintan** [comment]
@poush could you please suggest a workaround as there is no proper wrapper function for getting cookies .

File: libraries/Config.php (hunk 4 above)
Comment (nijel) [inline libraries/Config.php]: It's better to store within some singleton, most likely Response class. This way you can also properly handle setting them once when sending response to client.

File: libraries/Config.php [hunk 7]
```diff
@@ -1868,8 +1875,81 @@ public function setCookie($cookie, $value, $default = null,
...
     }
+    /**
+     * sets the cookie
+     * @param string      $cookieArr Cookie array name
+     * @param string      $cookie    sub cookie name
+     * @param mixed       $value     sub cookie value
+     * @param int         $validity validity of cookie in seconds (default is one month)
+     * @param bool        $httponly whether cookie is only for HTTP (and not for scripts)
+     */
+    public function setCookieArr($cookieArr,$cookie, $value,
```
Comment (nijel) [inline libraries/Config.php]: Thinking more about this, the cookie handling really doesn't belong to the Config class, but rather to Response one.

Comment (zaverichintan) [inline libraries/Config.php]: Should I change the file for the function? As I too feel the same. It should be with other wrapper  functions. @nijel  kindly suggest the file.

**Comment by bhagyeshworld** [comment]
@zaverichintan You are missing class object reference while calling getCookieArr() in /home/travis/build/phpmyadmin/phpmyadmin/libraries/plugins/auth/AuthenticationCookie.php on line 412.

Try calling that functoin using $GLOBALS['PMA_Config']->setCookie(... and it should build well !!

**Pull request closed without merging**